lib/Sema/TypeCheckAttr.cpp

```diff
@@ -4142,46 +4142,219 @@ static bool typeCheckDerivativeAttr(ASTContext &Ctx, Decl *D,
   auto *derivative = cast<FuncDecl>(D);
   auto lookupConformance =
       LookUpConformanceInModule(D->getDeclContext()->getParentModule());
-  auto originalName = attr->getOriginalFunctionName();
+
+  // Resolve the referenced original function declaration.
+  auto *originalAFD = evaluateOrDefault(
+      Ctx.evaluator, DerivativeAttrOriginalDeclRequest{attr, derivative},
+      nullptr);
+  if (!originalAFD)
+    return true;
 
   auto *derivativeInterfaceType =
       derivative->getInterfaceType()->castTo<AnyFunctionType>();
-
-  // Perform preliminary `@derivative` declaration checks.
-  // The result type should be a two-element tuple.
-  // Either a value and pullback:
-  //     (value: R, pullback: (R.TangentVector) -> (T.TangentVector...)
-  // Or a value and differential:
-  //     (value: R, differential: (T.TangentVector...) -> (R.TangentVector)
-  auto derivativeResultType = derivative->getResultInterfaceType();
+  auto *originalFnType =
+      originalAFD->getInterfaceType()->castTo<AnyFunctionType>();
+  auto derivativeResultType =
+      evaluateOrDefault(Ctx.evaluator, ResultTypeRequest{derivative}, Type());
   auto derivativeResultTupleType = derivativeResultType->getAs<TupleType>();
-  if (!derivativeResultTupleType ||
-      derivativeResultTupleType->getNumElements() != 2) {
-    diags.diagnose(attr->getLocation(),
-                   diag::derivative_attr_expected_result_tuple);
-    return true;
-  }
+  assert(derivativeResultTupleType->getNumElements() == 2);
   auto valueResultElt = derivativeResultTupleType->getElement(0);
   auto funcResultElt = derivativeResultTupleType->getElement(1);
-  // Get derivative kind and derivative function identifier.
-  AutoDiffDerivativeFunctionKind kind;
+
+  // Resolve the derivative function kind;
+  AutoDiffDerivativeFunctionKind derivativeKind;
   if (valueResultElt.getName().str() != "value") {
     diags.diagnose(attr->getLocation(),
                    diag::derivative_attr_invalid_result_tuple_value_label);
     return true;
   }
   if (funcResultElt.getName().str() == "differential") {
-    kind = AutoDiffDerivativeFunctionKind::JVP;
+    derivativeKind = AutoDiffDerivativeFunctionKind::JVP;
   } else if (funcResultElt.getName().str() == "pullback") {
-    kind = AutoDiffDerivativeFunctionKind::VJP;
+    derivativeKind = AutoDiffDerivativeFunctionKind::VJP;
   } else {
     diags.diagnose(attr->getLocation(),
                    diag::derivative_attr_invalid_result_tuple_func_label);
     return true;
   }
-  attr->setDerivativeKind(kind);
+  attr->setDerivativeKind(derivativeKind);
```

Comment:
This state should be next on the chopping block. A nice thing to do would be adding another accessor that forces `DerivativeAttrOriginalDeclRequest` and computes this from it.